src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreReference.cs

```diff
@@ -14,21 +19,52 @@ public class ProjectRestoreReference
         /// </summary>
         public string ProjectPath { get; set; }
 
+        public LibraryIncludeFlags IncludeAssets { get; set; } = LibraryIncludeFlags.All;
+
+        public LibraryIncludeFlags ExcludeAssets { get; set; }
+
+        public LibraryIncludeFlags PrivateAssets { get; set; } = LibraryIncludeFlagUtils.DefaultSuppressParent;
+
         public override int GetHashCode()
         {
-            return StringComparer.Ordinal.GetHashCode(ProjectUniqueName);
+            var combiner = new HashCodeCombiner();
+
+            combiner.AddInt32(StringComparer.Ordinal.GetHashCode(ProjectPath));
+            combiner.AddInt32(StringComparer.OrdinalIgnoreCase.GetHashCode(ProjectUniqueName));
+            combiner.AddObject(IncludeAssets);
+            combiner.AddObject(ExcludeAssets);
```

Comment:
The flags are ushort so it would not hit int, it could be optimized but there shouldn't be enough comparisons on this type to make it worthwhile.